File: js/t1/tests/model.fuzz.test.js
import { describe, test } from 'node:test';
import assert from 'node:assert/strict';
import { T } from '../model.js';
import * as TI from './testInstance.js';

const eng = T(TI.Piece, TI.InitialMainGrid, TI.ForbiddenGrid, TI.RotGrid, TI.InitialY, TI.InitialX,
              TI.PW, TI.FY, TI.FX);

function randOf(arr) { return arr[Math.floor(Math.random() * arr.length)]; }

function randomTrace(steps) {
  const m = new eng.Machine(randOf(TI.Piece));
  let sawGameover = false;
  for (let i = 0; i < steps; i++) {
    const kind = Math.floor(Math.random() * 4);
    assert.doesNotThrow(() => {
      if (kind === 0) m.movePiece(randOf([-1, 0]), randOf([-1, 0, 1]));
      else if (kind === 1) m.rotatePiece(Math.random() < 0.5);
      else if (kind === 2) m.fixPiece(randOf(TI.Piece));
      else m.fallStep(randOf(TI.Piece));
    });

    assert.strictEqual(eng.typeOK(m), true);
    for (const f of [m.py, m.px, m.pr, m.clearedLines])
      assert.ok(Math.abs(f) <= Number.MAX_SAFE_INTEGER);
    if (sawGameover) assert.strictEqual(m.gameover, true);
    if (m.gameover) sawGameover = true;
  }
}

describe('T1 model.js fuzz', () => {
  test('long random traces over the test instance: no throw + typeOK + bounds + gameover-monotone', () => {
    for (let trial = 0; trial < 20; trial++) randomTrace(2000);
  });

  test('adversarial checkAxioms: malformed params trip an assertion', () => {
    const originalAssert = console.assert;
    let tripped = false;
    console.assert = (cond) => { if (!cond) tripped = true; };
    try {
      // ragged grid
      tripped = false;
      const raggedGrid = [[false, false], [false]];
      const badEng = T(TI.Piece, raggedGrid, TI.ForbiddenGrid, TI.RotGrid, TI.InitialY, TI.InitialX,
                        TI.PW, TI.FY, TI.FX);
      badEng.checkAxioms();
      assert.strictEqual(tripped, true);

      // non-primitive piece
      tripped = false;
      const badEng2 = T([{}], TI.InitialMainGrid, TI.ForbiddenGrid, TI.RotGrid, () => 4, () => 1,
                         TI.PW, TI.FY, TI.FX);
      badEng2.checkAxioms();
      assert.strictEqual(tripped, true);

      // B > MAX_SAFE_INTEGER (force via a tiny MAX_SAFE_INTEGER override)
      tripped = false;
      const badEng3 = T(TI.Piece, TI.InitialMainGrid, TI.ForbiddenGrid, TI.RotGrid, TI.InitialY, TI.InitialX,
                         TI.PW, TI.FY, TI.FX, 1 /* MAX_SAFE_INTEGER */);
      badEng3.checkAxioms();
      assert.strictEqual(tripped, true);

      // negative FY
      tripped = false;
      const badEng4 = T(TI.Piece, TI.InitialMainGrid, TI.ForbiddenGrid, TI.RotGrid, TI.InitialY, TI.InitialX,
                         TI.PW, -1, TI.FX);
      badEng4.checkAxioms();
      assert.strictEqual(tripped, true);
    } finally {
      console.assert = originalAssert;
    }
  });
});

File: rs/t1/tests/model_fuzz_test.rs
//! spec: tests/model_fuzz_test.rs — §9.3, `implementation.md`.

#[path = "test_instance.rs"]
mod fixture;

use fixture::{Prng, TestInstance};
use t1::instance::Tetris;
use t1::model::{self, Machine, Params};

/// Runs `steps` random commands against a fresh `Machine<P>`, asserting no
/// panic, `type_ok`, integer-field bounds (D10), and gameover monotonicity
/// after every single step. Does not check the other four `Correct`
/// conjuncts (would just be re-checking `model.rs` against itself);
/// `tests/oracle.rs` (§9.4) verifies those independently at lower step counts.
fn fuzz<P: Params>(seed: u64, steps: usize) {
    let hm = P::initial_main_grid().len() as i64;
    let wm = if hm > 0 {
        P::initial_main_grid()[0].len() as i64
    } else {
        0
    };
    let b = std::cmp::max(hm, wm) + P::PW - 1;

    let mut rng = Prng::new(seed);
    let pieces = P::piece_all();
    let start = rng.choose(pieces);
    let mut m = Machine::<P>::new(start, || rng.choose(pieces));
    let mut prev_gameover = m.gameover;

    for step in 0..steps {
        match rng.next_below(7) {
            0 => {
                m.move_piece(0, -1);
            }
            1 => {
                m.move_piece(0, 1);
            }
            2 => {
                m.move_piece(-1, 0);
            }
            3 => {
                m.rotate_piece(true);
            }
            4 => {
                m.rotate_piece(false);
            }
            5 => {
                m.fix_piece(rng.choose(pieces));
            }
            _ => {
                m.fall_step(rng.choose(pieces));
            }
        }

        assert!(
            model::type_ok(&m),
            "type_ok failed at step {step}: {:?}",
            m.mg
        );
        assert!(m.py.abs() <= b, "py out of bounds at step {step}: {}", m.py);
        assert!(m.px.abs() <= b, "px out of bounds at step {step}: {}", m.px);
        assert!(
            !(prev_gameover && !m.gameover),
            "gameover flipped false at step {step}"
        );
        prev_gameover = m.gameover;
    }
}

#[test]
fn fuzz_test_instance_long_trace() {
    fuzz::<TestInstance>(0x1234_5678_9abc_def0, 20_000);
}

#[test]
fn fuzz_test_instance_many_seeds() {
    // shorter independent traces to catch seed-dependent issues a single run might miss
    for seed in 1u64..=20 {
        fuzz::<TestInstance>(seed, 2_000);
    }
}

#[test]
fn fuzz_real_instance_long_trace() {
    fuzz::<Tetris>(0xdead_beef_cafe_f00d, 5_000);
}

// ── Adversarial check_axioms ─────────────────────────────────────────────
//
// Violates `AxiomsPW` (`PW = 0`). `check_axioms` checks `AxiomsPW` first, so
// `unreachable!()` bodies suffice for the rest of this impl.

struct MalformedPw;

impl Params for MalformedPw {
    type Piece = ();
    type CellExtra = std::convert::Infallible;

    const PW: i64 = 0; // violates AxiomsPW (`PW > 0`)
    const FY: i64 = 0;
    const FX: i64 = 0;

    fn piece_all() -> &'static [()] {
        unreachable!("check_axioms fails on AxiomsPW before reaching this")
    }
    fn initial_main_grid() -> &'static Vec<Vec<bool>> {
        unreachable!()
    }
    fn forbidden_grid() -> &'static Vec<Vec<bool>> {
        unreachable!()
    }
    fn rot_grid(_p: (), _r: u8) -> &'static Vec<Vec<Option<()>>> {
        unreachable!()
    }
    fn initial_y(_p: ()) -> i64 {
        unreachable!()
    }
    fn initial_x(_p: ()) -> i64 {
        unreachable!()
    }
}

#[test]
#[should_panic]
fn malformed_pw_trips_check_axioms() {
    model::check_axioms::<MalformedPw>();
}

/// Violates `AxiomsInitialYX` (spawn position outside the main grid).
/// Reachable only after `AxiomsPW`/dimension checks pass, so the rest of
/// this impl must be genuinely well-formed.
struct MalformedInitialYX;

const MALFORMED_MAIN_GRID: [[bool; 3]; 3] = [[false; 3]; 3];
const MALFORMED_FORBIDDEN_GRID: [[bool; 3]; 1] = [[true; 3]];
const MALFORMED_PIECE_GRID: [[Option<()>; 3]; 3] = [
    [None, None, None],
    [Some(()), Some(()), Some(())],
    [None, None, None],
];

impl Params for MalformedInitialYX {
    type Piece = ();
    type CellExtra = std::convert::Infallible;

    const PW: i64 = 3;
    const FY: i64 = 2;
    const FX: i64 = 0;

    fn piece_all() -> &'static [()] {
        &[()]
    }
    fn initial_main_grid() -> &'static Vec<Vec<bool>> {
        static G: std::sync::LazyLock<Vec<Vec<bool>>> =
            std::sync::LazyLock::new(|| MALFORMED_MAIN_GRID.iter().map(|r| r.to_vec()).collect());
        &G
    }
    fn forbidden_grid() -> &'static Vec<Vec<bool>> {
        static G: std::sync::LazyLock<Vec<Vec<bool>>> = std::sync::LazyLock::new(|| {
            MALFORMED_FORBIDDEN_GRID
                .iter()
                .map(|r| r.to_vec())
                .collect()
        });
        &G
    }
    fn rot_grid(_p: (), _r: u8) -> &'static Vec<Vec<Option<()>>> {
        static G: std::sync::LazyLock<Vec<Vec<Option<()>>>> =
            std::sync::LazyLock::new(|| MALFORMED_PIECE_GRID.iter().map(|r| r.to_vec()).collect());
        &G
    }
    fn initial_y(_p: ()) -> i64 {
        1000 // far outside the 3-row main grid — violates AxiomsInitialYX
    }
    fn initial_x(_p: ()) -> i64 {
        0
    }
}

#[test]
#[should_panic]
fn malformed_initial_y_trips_check_axioms() {
    model::check_axioms::<MalformedInitialYX>();
}
